aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1892,36 +1935,146 @@ TORCH_IMPL_FUNC(bmm_out_cpu)
+    TORCH_CHECK(
+        (output_device == input1_device) && (input1_device == input2_device),
+        "vdot: Expected the output and input tensors to be on the "
+        "same device, but got the output tensor on ",
+        output_device,
+        ", the 'input' tensor on ",
+        input1_device,
+        ", and the 'other' tensor on ",
+        input2_device);
+    if (self.is_complex()) {
+    if (self.is_conj()) {
+      if (other.is_conj()) {
+        Tensor temp_result = at::empty({}, result.options());
+        dot_out_impl(self.conj(), other.conj(), temp_result);
+        result.copy_(temp_result.conj());
+      } else {
+        vdot_out_impl(other.conj(), self, result);
+      }
+      return;
+    } else if (other.is_conj()) {
+      vdot_out_impl(other.conj(), self, result);
+      return;
+    }
+    }
+
+    if (self._is_zerotensor() || other._is_zerotensor()) {
+    result.fill_(0);
+    }
+
+    if (use_mkldnn_matmul(self, other, /*result=*/Tensor())) {
+    // mkldnn matmul expect result have sizes info to create ideep tensor
+    auto r = at::empty({1, 1}, self.options());
+    mkldnn_matmul(self, other, r, /*beta=*/0);
+    result.fill_(r);
+    return;
+    }
+
+    return AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(
+        at::ScalarType::BFloat16,
+        at::ScalarType::Half,
+        self.scalar_type(),
+        "dot",
+        [&] {
+          result.fill_(at::native::dot_impl<scalar_t>(
+              self.numel(),
+              const_cast<scalar_t*>(self.const_data_ptr<scalar_t>()),
+              self.stride(0),
+              const_cast<scalar_t*>(other.const_data_ptr<scalar_t>()),
+              other.stride(0)));
+        });
+}
+
+void vdot_out_impl(
+    const Tensor& self,
+    const Tensor& other,
+    const Tensor& result) {
+    auto output_device = result.device();
+    auto input1_device = self.device();
+    auto input2_device = other.device();
+    // check if the input & output tensors are on the same device.
+    TORCH_CHECK(
+        (output_device == input1_device) && (input1_device == input2_device),
+        "vdot: Expected the output and input tensors to be on the "
+        "same device, but got the output tensor on ",
+        output_device,
+        ", the 'input' tensor on ",
+        input1_device,
+        ", and the 'other' tensor on ",
+        input2_device);
+    // Dispatch to `dot` for real dtypes.
+    if (!self.is_complex()) {
+    dot_out_impl(self, other, result);
+    return;
+    }
+
+    if (self.is_conj()) {
+    if (other.is_conj()) {
+      vdot_out_impl(other.conj(), self.conj(), result);
+    } else {
+      dot_out_impl(self.conj(), other, result);
+    }
+    return;
+    } else if (other.is_conj()) {
+    Tensor temp_result = at::empty({}, result.options());
+    dot_out_impl(self, other.conj(), result);
+    result.copy_(temp_result.conj());
+    return;
+    }
+
+    if (self._is_zerotensor() || other._is_zerotensor()) {
+    result.fill_(0);
```

Comment:
The behavior is changed - before, this path would use ZeroTensor allocator: https://github.com/pytorch/pytorch/blob/ff616c26fb66b26966b6f6f50aa85bf615400816/aten/src/ATen/native/TensorFactories.h#L119-L120, which seems to not fill result with 0.